docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,292 @@
+     - **Challenges:**
+       - **Unintended State Changes:** Increasing the ETCD replicas to bring the cluster back up modifies the cluster's state without the operator's explicit intention, which could lead to unexpected behavior or conflicts with operational policies.
+       - **Operational Overhead:** Automating the process of scaling the ETCD cluster up and down adds complexity and potential risks, especially if the scaling operations interfere with other scheduled tasks or maintenance windows.
+
+   - **Option B:** Start an embedded ETCD instance and take a snapshot.
+     - **Advantages:**
+       - **Respects Operator Intentions:** Does not alter the state of the actual ETCD cluster, keeping it in hibernation as intended by the operator.
+       - **No Impact on Existing Setup:** Avoids modifying the ETCD cluster's replicas, preventing any unintended consequences.
+       - **Automated Process:** Can be integrated into existing maintenance jobs without manual intervention.
+
+**Proposed Solution:**
+
+*Option B is recommended* because it respects the operator's intention to keep the ETCD cluster hibernated and avoids the complexities and risks associated with modifying the cluster's state.
+
+Leverage the compaction job to take fresh snapshots periodically during hibernation. This ensures that:
+
+- **Immutability is Maintained:** New snapshots have a fresh immutability period, keeping backups protected.
+- **Operator Intentions are Respected:** The ETCD cluster remains in its hibernated state, as per the operator's intention.
+- **Storage Costs are Controlled:** Avoids unnecessary extension of immutability for all snapshots, preventing increased storage costs.
+
+### New Compaction Controller Flag
+
+**Flag:** `--hibernation-snapshot-interval`
+
+- **Type:** Duration
+- **Default:** `24h`
+- **Description:** This flag sets the period after which a compaction job should be triggered for a hibernated ETCD cluster, based on the time since the last renewal of the full snapshot lease. If the time since `fullLease.Spec.RenewTime.Time` exceeds the duration specified by this flag, and `etcd.spec.replicas` is `0` (indicating hibernation), the compaction job will automatically trigger to create a new snapshot. This approach ensures that backups remain within the immutability period and are safeguarded against becoming mutable.
+
+### Garbage Collection during Compaction for Hibernated ETCD Clusters
+
+If an ETCD cluster is hibernated for a long duration, there is a chance that backup storage will accumulate a large number of snapshots. Since the ETCD cluster is not running, the `etcd-backup-restore` component is also not running to perform garbage collection of the snapshots.
+
+To address this, the compaction job should be enhanced to handle garbage collection during hibernation. This ensures that old snapshots are cleaned up appropriately, considering the immutability constraints.
```

Comment:
will this be done during regular compaction job or with the one created for hibernated clusters. Again because the word compaction is overloaded as we will not be compacting anything for hibernated clusters except taking a full snapshot again.

Reply:
A separate command will be introduced in etcd-backup-restore for garbage collection.

This will be added to the compaction job approach.